nx/lib/nx/defn/expr.ex

```diff
@@ -984,9 +984,16 @@ defmodule Nx.Defn.Expr do
 
   defp constant(%{type: type, shape: shape} = out, number) do
     number =
-      if is_integer(number) and Nx.Type.float?(type),
-        do: Complex.multiply(1.0, number),
-        else: number
+      cond do
+        is_integer(number) and Nx.Type.float?(type) ->
+          Complex.multiply(1.0, number)
+
+        is_float(number) and Nx.Type.integer?(type) ->
+          floor(number)
```

Comment:
I don't think we should do this here, it most often is a bug. I think here we should raise and we do explicit conversions on as_type.

Reply:
I think it's a valid case, see

https://github.com/elixir-nx/nx/blob/1f4ddfea59849da1291083cac51177915511d0f5/nx/lib/nx/defn/expr.ex#L586-L587

This is an optimisation when casting constants, so we need to cast the actual value accordingly.

Reply:
Yes, the bug is in `as_type`. We can cast from integer -> float, but never from float to integer transparently, unless using `as_type`. So the logic belongs there! :)